### PR TITLE
feat: add support for model inheritance

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

This PR makes extending a model inherit the relevant model options, attributes, and associatiosn

This feature is only available for the new decorator-based model declaration.

## Abstract Models

the `@Table.Abstract` decorator can be used to declare abstract models that can only be used for inheritance.

These models are never initialized. They only exist to register metadata on them.

## Model Option Inheritance

Options are inherited from parent models just like if you used the `@Table` decorator multiple times on the final model. In case of conflicting options, the options set on the descendant overwrite the options set on the parent instead of throwing an error.

The `tableName`, `name` and `modelName` options are never inherited.

## Attribute Inheritance

Options are inherited from parent models just like if you used the `@Attribute` decorator multiple times on the final model. In case of conflicting options, the options set on the descendant overwrite the options set on the parent instead of throwing an error.

## Association Inheritance

Only `@BelongsTo` _without_ the `inverse` option can be inherited. All other association types, or using the `inverse` option will throw an error.

This limitation is put into place to prevent associations from declare conflicting associations on the target model.

This could be revisited in the future by providing a way to apply different options based on which child model is being initialized, like this:

```ts
class BaseUser extends Model<InferAttributes<BaseUser>> {
  @hasMany(() => Project, SourceModel => ({
    // this will declare a different foreign key on `Project` for each class extending BaseUser
    foreignKey: lowerFirst(SourceModel.name) + 'Id',
    inverse: {
      as: lowerFirst(SourceModel.name),
    },
  }))
  declare projects?: NonAttribute<Project[]>;
}
```

But that is currently out of scope.

## Todos

- Document model inheritance in documentation repository

### CODE REVIEW

File: packages/core/test/unit/decorators/table.test.ts [hunk 1]
```diff
@@ -121,4 +122,320 @@ describe(`@Table legacy decorator`, () => {
       return User;
     }).to.throw();
   });
+
+  it('is inheritable', () => {
```
Comment (WikiRik) [inline packages/core/test/unit/decorators/table.test.ts:126]: Nice explanations!

File: packages/core/test/unit/decorators/table.test.ts [hunk 2]
```diff
@@ -121,4 +122,320 @@ describe(`@Table legacy decorator`, () => {
...
+
+    // must be the same value but a different instance (cloned)
+    expect(baseOptions.scopes).not.to.equal(inheritedOptions.scopes, 'scopes option must be a different instance');
+    expect(baseOptions.defaultScope).not.to.equal(inheritedOptions.defaultScope, 'defaultScope option must be a different instance');
+    expect(baseOptions.indexes).not.to.equal(inheritedOptions.indexes, 'indexes option must be a different instance');
+    expect(baseOptions.hooks).not.to.equal(inheritedOptions.hooks, 'indexes option must be a different instance');
+    expect(baseOptions.validate).not.to.equal(inheritedOptions.validate, 'validate option must be a different instance');
+  });
+
+  it('overwrites defaultScope', () => {});
```
Comment (WikiRik) [inline packages/core/test/unit/decorators/table.test.ts:317]: This test should not be empty, right?

**Review by WikiRik — COMMENTED** [review]
One compliment and one small question, but apart from that it looks great!

File: packages/core/test/unit/decorators/table.test.ts (hunk 2 above)
Comment (ephys) [reply packages/core/test/unit/decorators/table.test.ts:317]: Woops

Comment (ephys) [reply packages/core/test/unit/decorators/table.test.ts]: Fixed

**Review by WikiRik — APPROVED** [review]

**Pull request merged**